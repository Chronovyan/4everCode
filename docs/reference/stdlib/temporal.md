---
title: Temporal Module
description: Time and timeline manipulation
weight: 10
---

# Temporal Module

## Overview

The `std.temporal` module provides primitives for time manipulation, timeline management, and temporal programming in Chronovyan.

## Timeline Management

### Types

#### `Timeline<T>`
Represents a sequence of temporal events.

```chronovyan
// Create a new timeline
DECLARE timeline = TIMELINE_NEW("main");
```

### Functions

#### `TIMELINE_NEW(name: STRING) -> Timeline`
Creates a new timeline with the given name.

#### `TIMELINE_ADD_EVENT(timeline: Timeline, when: MOMENT, callback: FUNCTION)`
Schedules an event to occur at a specific time.

```chronovyan
TIMELINE_ADD_EVENT(timeline, NOW() + 1s, \() {
    PRINT("One second has passed");
});
```

#### `TIMELINE_BRANCH(timeline: Timeline, name: STRING) -> Timeline`
Creates a new branch from an existing timeline.

#### `TIMELINE_MERGE(source: Timeline, target: Timeline)`
Merges the source timeline into the target timeline.

## Temporal Variables

### `CONF` (Conformity)
Immutable variables that maintain temporal consistency.

```chronovyan
DECLARE CONF::STATIC VERSION = "1.0.0";
```

### `REB` (Rebellion)
Mutable variables that can change over time.

```chronovyan
DECLARE REB::FLUX counter = 0;
```

## Quantum Operations

### `QUANTUM<T>`
Represents a quantum state that can be in superposition.

```chronovyan
DECLARE q = QUBIT(0);
H(q);  // Put into superposition
```

### Common Quantum Gates

- `H(qubit)`: Hadamard gate
- `X(qubit)`: Pauli-X gate
- `CNOT(control, target)`: Controlled-NOT gate
- `MEASURE(qubit) -> BOOL`: Measures a qubit

## Temporal Resources

### `AETHEL`
Temporal energy resource used for temporal operations.

```chronovyan
DECLARE energy: AETHEL = 100.0;
```

### `CHRONON`
Basic quantum of time.

```chronovyan
DECLARE start = NOW();
// ...
DECLARE elapsed = NOW() - start;
```

## Example: Simple Timeline

```chronovyan
IMPORT std.temporal as time;

FUNC VOID main() {
    DECLARE tl = time.TIMELINE_NEW("example");
    
    // Schedule events
    FOR i IN 1..5 {
        time.TIMELINE_ADD_EVENT(tl, time.NOW() + i * 1s, \() {
            PRINT("Event ", i, " at ", time.NOW());
        });
    }
    
    // Run the timeline
    time.TIMELINE_RUN(tl);
}
```

## Performance Considerations

- Timeline operations are O(1) for insertion and O(log n) for event dispatch
- Quantum operations require significant computational resources
- Use `CONF` variables for values that don't change to enable optimizations

## Related

- [Concurrency](./concurrency.md) - For concurrent programming with temporal types
- [Time](./time.md) - For date and time handling
- [Random](./random.md) - For generating random temporal values
- [Testing](./testing.md) - For testing temporal code

## Examples

### Timeline Visualization

```mermaid
timeline
    title Project Timeline
    section Phase 1
        Task 1 :a1, 2023-01-01, 30d
        Task 2 :after a1, 20d
    section Phase 2
        Task 3 :2023-03-01, 12d
        Task 4 :5d after Task 3
```

### Quantum Circuit

```mermaid
quantumcircuit
    qreg q[2];
    creg c[2];
    
    h q[0];
    cx q[0], q[1];
    measure q[0] -> c[0];
    measure q[1] -> c[1];
```

## See Also

- [Standard Library](../stdlib/index.md) - Overview of all standard library modules
- [Chronovyan Language Specification](reference/language/specification.md) - For language-level temporal features
