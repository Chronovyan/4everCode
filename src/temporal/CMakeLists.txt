# Add the timeline library
add_library(timeline
    timeline.cpp
)

target_include_directories(timeline
    PUBLIC 
        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../../include>
        $<INSTALL_INTERFACE:include>
    PRIVATE
        ${CMAKE_CURRENT_SOURCE_DIR}
)

target_link_libraries(timeline
    PRIVATE
        Threads::Threads
)

# Install the library
install(TARGETS timeline
    EXPORT ChronovyanTargets
    LIBRARY DESTINATION lib
    ARCHIVE DESTINATION lib
    RUNTIME DESTINATION bin
)

# Add tests if enabled
if(BUILD_TESTS)
    add_executable(timeline_test
        ${CMAKE_CURRENT_SOURCE_DIR}/../../test/timeline_test.cpp
    )
    
    target_link_libraries(timeline_test
        PRIVATE
            timeline
            gtest
            gtest_main
            Threads::Threads
    )
    
    # Enable CTest
    enable_testing()
    
    # Add the test to CTest
    add_test(NAME timeline_test
        COMMAND timeline_test
    )
    
    # Create a custom target for building tests
    add_custom_target(run_timeline_tests
        COMMAND ctest --output-on-failure
        WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
        DEPENDS timeline_test
    )
    
    # Create a custom target to run tests
    add_custom_target(run_tests
        DEPENDS run_timeline_tests
        COMMENT "Running timeline tests..."
    )
endif()
