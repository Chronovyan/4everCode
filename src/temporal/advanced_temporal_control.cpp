#include "advanced_temporal_control.h"

namespace chronovyan {

// Stub implementation for the advanced temporal control
// Will be implemented in future updates

} // namespace chronovyan