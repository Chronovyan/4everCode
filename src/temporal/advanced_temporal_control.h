#pragma once

namespace chronovyan {

// Stub header for the advanced temporal control
// Will be implemented in future updates

}  // namespace chronovyan