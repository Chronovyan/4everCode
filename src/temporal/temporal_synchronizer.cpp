#include "temporal_synchronizer.h"

namespace chronovyan {

// Stub implementation for the temporal synchronizer
// Will be implemented in future updates

} // namespace chronovyan