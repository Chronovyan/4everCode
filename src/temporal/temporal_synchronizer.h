#pragma once

namespace chronovyan {

// Stub header for the temporal synchronizer
// Will be implemented in future updates

} // namespace chronovyan