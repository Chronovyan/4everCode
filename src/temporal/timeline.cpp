#include <algorithm>
#include <chronovyan/timeline.hpp>
#include <shared_mutex>
#include <stdexcept>

namespace chronovyan {
namespace sync {

void Timeline::add_sync_point(const SyncPoint& point) {
    std::unique_lock<std::shared_mutex> lock(mutex_);
    sync_points_.push_back(point);

    // Sort by timestamp to maintain chronological order
    std::sort(sync_points_.begin(), sync_points_.end(),
              [](const SyncPoint& a, const SyncPoint& b) { return a.timestamp < b.timestamp; });
}

SyncPoint Timeline::get_current_sync_point() const {
    std::shared_lock<std::shared_mutex> lock(mutex_);
    if (sync_points_.empty()) {
        throw std::runtime_error("No sync points available in timeline");
    }
    return sync_points_.back();
}

std::vector<SyncPoint> Timeline::get_recent_sync_points(size_t count) const {
    std::shared_lock<std::shared_mutex> lock(mutex_);
    if (sync_points_.empty()) {
        return {};
    }

    size_t start = sync_points_.size() > count ? sync_points_.size() - count : 0;
    return std::vector<SyncPoint>(sync_points_.begin() + start, sync_points_.end());
}

void Timeline::clear() {
    std::unique_lock<std::shared_mutex> lock(mutex_);
    sync_points_.clear();
}

size_t Timeline::size() const {
    std::shared_lock<std::shared_mutex> lock(mutex_);
    return sync_points_.size();
}

bool Timeline::empty() const {
    std::shared_lock<std::shared_mutex> lock(mutex_);
    return sync_points_.empty();
}

// TimelineManager implementation

Timeline& TimelineManager::get_timeline(const std::string& name) {
    std::unique_lock<std::shared_mutex> lock(mutex_);
    return timelines_[name];
}

bool TimelineManager::has_timeline(const std::string& name) const {
    std::shared_lock<std::shared_mutex> lock(mutex_);
    return timelines_.find(name) != timelines_.end();
}

void TimelineManager::remove_timeline(const std::string& name) {
    std::unique_lock<std::shared_mutex> lock(mutex_);
    timelines_.erase(name);
}

void TimelineManager::add_event(const TimelineEvent& event) {
    std::unique_lock<std::shared_mutex> lock(mutex_);
    event_log_.push_back(event);

    // Keep only the most recent 1000 events
    const size_t max_events = 1000;
    if (event_log_.size() > max_events) {
        event_log_.erase(event_log_.begin(), event_log_.begin() + (event_log_.size() - max_events));
    }
}

std::vector<TimelineEvent> TimelineManager::get_recent_events(size_t count) const {
    std::shared_lock<std::shared_mutex> lock(mutex_);
    if (event_log_.empty()) {
        return {};
    }

    size_t start = event_log_.size() > count ? event_log_.size() - count : 0;
    return std::vector<TimelineEvent>(event_log_.begin() + start, event_log_.end());
}

}  // namespace sync
}  // namespace chronovyan
