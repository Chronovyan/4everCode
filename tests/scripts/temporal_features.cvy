// Test of temporal features in Chronovyan

// Import temporal modules
import temporal;
import debt_tracker;

// Initialize a temporal runtime
var runtime = new TemporalRuntime();
print("Temporal runtime initialized");

// Create a TimeAnchor
var anchor = runtime.createAnchor("MainTimelineAnchor");
print("Created TimeAnchor: " + anchor.getName());

// Track temporal debt
var tracker = new TemporalDebtTracker();
print("Initial stability: " + tracker.getStability());

// Simulate some temporal operations
print("\nPerforming temporal operations:");

// Simulating a minor temporal operation (e.g., small loop)
tracker.recordOperation("MinorLoop", 1);
print("After minor operation - Stability: " + tracker.getStability());

// Simulating a moderate temporal operation (e.g., branch creation)
tracker.recordOperation("BranchCreate", 2);
print("After branch creation - Stability: " + tracker.getStability());

// Project paradox risk for a major operation
var riskLevel = tracker.projectParadoxRisk("MajorRewrite", 3, 2);
print("Projected risk for major rewrite: " + riskLevel);

// Creating a timestream and echo
var stream = runtime.createTimestream("ExperimentalBranch");
print("Created timestream: " + stream.getId());

var echo = stream.createEcho("TestEcho");
print("Created echo: " + echo.getId());

// Check status of system
print("\nFinal system status:");
print("Current stability: " + tracker.getStability());
print("Active timestreams: " + runtime.getActiveTimestreamCount());
print("Total anchors: " + runtime.getAnchorCount());

// Attempt harmonization
var harmonizationSuccess = runtime.harmonize(stream, anchor);
print("Harmonization " + (harmonizationSuccess ? "succeeded" : "failed"));

// Final debt assessment
print("Final temporal debt: " + tracker.getTotalDebt()); 