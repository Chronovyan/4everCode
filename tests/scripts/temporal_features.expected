Temporal runtime initialized
Created TimeAnchor: MainTimelineAnchor
Initial stability: 1.0

Performing temporal operations:
After minor operation - Stability: 0.95
After branch creation - Stability: 0.85
Projected risk for major rewrite: 0.42
Created timestream: ExperimentalBranch-TS001
Created echo: TestEcho-E001

Final system status:
Current stability: 0.85
Active timestreams: 1
Total anchors: 1
Harmonization succeeded
Final temporal debt: 0.15 